File: index_records_txt.ts
#!/usr/bin/env -S deno run -A
// deno run -A index_records_txt.ts --input ./epstein-justice-files-text/Datasets-9-12 --db records.sqlite --run content --dedupe 1 --chunk-size 2000
import { DB } from "https://deno.land/x/sqlite@v3.9.1/mod.ts";
import { bold, cyan, green, red, yellow } from "https://deno.land/std@0.224.0/fmt/colors.ts";
import { parse } from "https://deno.land/std@0.224.0/flags/mod.ts";
import { walk } from "https://deno.land/std@0.224.0/fs/walk.ts";

type Block = {
  sourceLine: string;
  metadataSource: string;
  metadataFilename: string;
  body: string;
};

type Stats = {
  docs: number;
  chunks: number;
  dupSkipped: number;
  lastFile?: string;
};

const SPLIT_MARKER = "----------------------------------------";
const SOURCE_RE = /^--- SOURCE:\s*(.+?)\s*---\s*$/gm;
const EFTA_RE = /\bEFTA\d{8}\b/;

function cleanText(text: string): string {
  return text.normalize("NFKC");
}

function sha256Hex(data: string): string {
  const bytes = new TextEncoder().encode(data);
  return [...new Uint8Array(crypto.subtle.digestSync("SHA-256", bytes))]
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

function parseBlocks(fileText: string): Block[] {
  const matches = [...fileText.matchAll(SOURCE_RE)];
  if (!matches.length) return [];

  const blocks: Block[] = [];
  for (let i = 0; i < matches.length; i += 1) {
    const match = matches[i];
    const start = match.index ?? 0;
    const end = i + 1 < matches.length ? (matches[i + 1].index ?? fileText.length) : fileText.length;
    const header = match[1] ?? "";
    const chunk = fileText.slice(start, end).trim();

    const metadataSource = chunk.match(/METADATA_SOURCE:\s*(.+)/)?.[1]?.trim() ?? "";
    const metadataFilename = chunk.match(/METADATA_FILENAME:\s*(.+)/)?.[1]?.trim() ?? "";

    const body = chunk.includes(SPLIT_MARKER)
      ? chunk.split(SPLIT_MARKER, 2)[1]?.trim() ?? ""
      : chunk.replace(/^--- SOURCE:.*---\s*/m, "").trim();

    blocks.push({
      sourceLine: header.trim(),
      metadataSource,
      metadataFilename,
      body,
    });
  }

  return blocks;
}

function extractEftaId(value: string): string | null {
  const match = value.match(EFTA_RE);
  return match?.[0] ?? null;
}

function guessType(body: string): string {
  const head = body.slice(0, 1500);
  const looksEmail = /(^|\n)\s*From:\s*/i.test(head) && /(^|\n)\s*Subject:\s*/i.test(head);
  return looksEmail ? "email_like" : "document_like";
}

function stripNoise(body: string): string {
  let t = body;
  t = t.replace(/--- PAGE \d+ ---/g, "\n");
  t = t.replace(/^METADATA_(SOURCE|FILENAME):.*$/gim, "");
  t = t.replace(/^Couldn't load plugin\.\s*$/gim, "");
  t = t.replace(/This communication contains information[\s\S]{0,900}?applicable law\.\s*/gi, "");
  t = t.replace(/please note[\s\S]{0,1400}?copyright -all rights reserved\s*/gi, "");
  t = t.replace(/\n{3,}/g, "\n\n").trim();
  return t;
}

function chunkParagraph(text: string, targetSize: number): string[] {
  const parts = text.split(/\n\s*\n+/).map((part) => part.trim()).filter(Boolean);
  if (!parts.length) return [text.trim()];

  const chunks: string[] = [];
  let current = "";
  for (const part of parts) {
    if (current && current.length + part.length + 2 > targetSize) {
      chunks.push(current);
      current = part;
    } else {
      current = current ? `${current}\n\n${part}` : part;
    }
  }
  if (current) chunks.push(current);
  return chunks;
}

function ensureSchema(db: DB): void {
  db.execute("PRAGMA journal_mode=WAL");
  db.execute("PRAGMA synchronous=NORMAL");
  db.execute("PRAGMA temp_store=MEMORY");
  db.execute("PRAGMA mmap_size=268435456");
  db.execute("PRAGMA busy_timeout=5000");

  db.execute(`
    CREATE TABLE IF NOT EXISTS docs(
      run_id TEXT NOT NULL,
      doc_id TEXT NOT NULL,
      meta_json TEXT,
      text_sha256 TEXT,
      text_chars INTEGER,
      PRIMARY KEY(run_id, doc_id)
    );
  `);

  db.execute(`
    CREATE TABLE IF NOT EXISTS chunks(
      id INTEGER PRIMARY KEY,
      uid TEXT NOT NULL UNIQUE,
      run_id TEXT NOT NULL,
      chunk_id TEXT NOT NULL,
      order_index INTEGER,
      doc_id TEXT,
      source_file TEXT,
      text TEXT
    );
  `);
  db.execute("CREATE INDEX IF NOT EXISTS idx_chunks_doc ON chunks(run_id, doc_id, order_index)");
  db.execute("CREATE INDEX IF NOT EXISTS idx_docs_hash ON docs(run_id, text_sha256)");
  db.execute("CREATE UNIQUE INDEX IF NOT EXISTS idx_chunks_uid ON chunks(uid)");

  db.execute(`
    CREATE VIRTUAL TABLE IF NOT EXISTS chunks_fts USING fts5(
      text,
      content='chunks',
      content_rowid='id',
      tokenize="unicode61"
    );
  `);
}

function formatBytes(bytes: number): string {
  const mb = bytes / (1024 * 1024);
  return `${mb.toFixed(2)} MB`;
}

function updateStatus(stats: Stats): void {
  const line = [
    `docs=${stats.docs.toLocaleString()}`,
    `chunks=${stats.chunks.toLocaleString()}`,
    `dup=${stats.dupSkipped.toLocaleString()}`,
    stats.lastFile ? `file=${stats.lastFile}` : "",
  ].filter(Boolean).join(" | ");
  Deno.stdout.writeSync(new TextEncoder().encode(`\r${line}   `));
}

function finishStatus(): void {
  Deno.stdout.writeSync(new TextEncoder().encode("\n"));
}

const flags = parse(Deno.args, {
  string: ["input", "db", "run"],
  boolean: ["dedupe", "reset"],
  default: {
    db: "records.sqlite",
    run: "content",
    dedupe: true,
    reset: false,
    "chunk-size": 2000,
  },
});

const inputDir = String(flags.input ?? "");
const dbPath = String(flags.db ?? "records.sqlite");
const runId = String(flags.run ?? "content").trim();
const dedupe = Boolean(flags.dedupe);
const chunkSize = Math.max(200, Number(flags["chunk-size"] ?? 2000));
const reset = Boolean(flags.reset);

if (!inputDir) {
  console.error(red("Error: --input is required"));
  Deno.exit(1);
}
if (!runId) {
  console.error(red("Error: --run must not be empty"));
  Deno.exit(1);
}

const inputStat = await Deno.stat(inputDir).catch(() => null);
if (!inputStat?.isDirectory) {
  console.error(red(`Error: input directory not found: ${inputDir}`));
  Deno.exit(1);
}

console.log(
  cyan(
    `runId="${runId}" | dedupe=${dedupe ? 1 : 0} | chunk_size=${chunkSize} | reset=${reset ? 1 : 0}`,
  ),
);

const db = new DB(dbPath);
ensureSchema(db);

if (reset) {
  console.log(yellow(`Resetting existing run_id data: ${runId}`));
  db.execute("BEGIN");
  db.query("DELETE FROM chunks_fts WHERE rowid IN (SELECT id FROM chunks WHERE run_id = ?)", [runId]);
  db.query("DELETE FROM chunks WHERE run_id = ?", [runId]);
  db.query("DELETE FROM docs WHERE run_id = ?", [runId]);
  db.execute("COMMIT");
}

const stmtDoc = db.prepareQuery(`
  INSERT INTO docs(run_id, doc_id, meta_json, text_sha256, text_chars)
  VALUES (?, ?, ?, ?, ?)
  ON CONFLICT(run_id, doc_id) DO UPDATE SET
    meta_json=excluded.meta_json,
    text_sha256=excluded.text_sha256,
    text_chars=excluded.text_chars;
`);

const stmtChunk = db.prepareQuery(`
  INSERT OR IGNORE INTO chunks(uid, run_id, chunk_id, order_index, doc_id, source_file, text)
  VALUES (?, ?, ?, ?, ?, ?, ?);
`);

const stmtFts = db.prepareQuery(`
  INSERT OR IGNORE INTO chunks_fts(rowid, text)
  VALUES (?, ?);
`);

const stmtDup = dedupe
  ? db.prepareQuery("SELECT doc_id FROM docs WHERE run_id = ? AND text_sha256 = ? LIMIT 1")
  : null;

const stats: Stats = { docs: 0, chunks: 0, dupSkipped: 0 };
let docSinceCommit = 0;

const encoder = new TextEncoder();

for await (const entry of walk(inputDir, { includeDirs: false, exts: [".txt"] })) {
  stats.lastFile = entry.path.split("/").pop() ?? entry.path;
  updateStatus(stats);

  const raw = await Deno.readTextFile(entry.path).catch(() => "");
  if (!raw) continue;

  const blocks = parseBlocks(cleanText(raw));
  if (!blocks.length) continue;

  db.execute("BEGIN");
  for (const block of blocks) {
    const efta =
      extractEftaId(block.sourceLine) ??
      extractEftaId(block.metadataFilename) ??
      extractEftaId(block.body);

    const docId = efta ?? `doc_${sha256Hex(`${block.sourceLine}|${entry.path}`)}`;
    const bodyClean = stripNoise(cleanText(block.body));
    const textSha = sha256Hex(bodyClean);

    if (stmtDup) {
      const dupRow = stmtDup.first([runId, textSha]);
      if (dupRow) {
        stats.dupSkipped += 1;
        continue;
      }
    }

    const meta = {
      doc_id: docId,
      type: guessType(bodyClean),
      source_txt_file: entry.path,
      record_source: block.sourceLine,
      metadata_source: block.metadataSource,
      metadata_filename: block.metadataFilename,
      text_sha256: textSha,
      chars: bodyClean.length,
      chunk_mode: "paragraph",
      chunk_target_size: chunkSize,
    };

    stmtDoc.execute([runId, docId, JSON.stringify(meta), textSha, bodyClean.length]);
    stats.docs += 1;
    docSinceCommit += 1;

    const parts = chunkParagraph(bodyClean, chunkSize);
    for (let i = 0; i < parts.length; i += 1) {
      const chunkId = `${docId}:${String(i).padStart(4, "0")}`;
      const uid = `${runId}:${chunkId}`;
      stmtChunk.execute([uid, runId, chunkId, i, docId, entry.path, parts[i]]);
      const rowid = db.lastInsertRowId;
      if (rowid) {
        stmtFts.execute([rowid, parts[i]]);
      }
      stats.chunks += 1;
    }
  }
  db.execute("COMMIT");

  if (docSinceCommit >= 200) {
    docSinceCommit = 0;
    db.execute("PRAGMA wal_checkpoint(PASSIVE)");
  }
}

finishStatus();
console.log(green("Indexing complete."));

db.execute("PRAGMA wal_checkpoint(TRUNCATE)");
try {
  db.execute("INSERT INTO chunks_fts(chunks_fts) VALUES('optimize')");
} catch {
  console.warn(yellow("FTS optimize skipped."));
}
db.execute("PRAGMA optimize");
db.execute("ANALYZE");
db.execute("VACUUM");

const docsCount = db.queryEntries<{ count: number }>(
  "SELECT COUNT(*) as count FROM docs WHERE run_id = ?",
  [runId],
)[0]?.count ?? 0;
const chunksCount = db.queryEntries<{ count: number }>(
  "SELECT COUNT(*) as count FROM chunks WHERE run_id = ?",
  [runId],
)[0]?.count ?? 0;
const avgChunk = db.queryEntries<{ avg: number }>(
  "SELECT AVG(LENGTH(text)) as avg FROM chunks WHERE run_id = ?",
  [runId],
)[0]?.avg ?? 0;

const dbStat = await Deno.stat(dbPath);

console.log(bold("============================================================"));
console.log(`Documents:           ${docsCount.toLocaleString()}`);
console.log(`Chunks:              ${chunksCount.toLocaleString()}`);
console.log(`Duplicates skipped:  ${stats.dupSkipped.toLocaleString()}`);
console.log(`Avg chunk size:      ${Math.round(avgChunk)} chars`);
console.log(`Database size:       ${formatBytes(dbStat.size)}`);
console.log(`Database file:       ${dbPath}`);
console.log(bold("============================================================"));

db.close();

File: index_records_txt.py
#!/usr/bin/env python3
# python index_records_txt.py --input ./epstein-justice-files-text/Datasets-9-12 --db records.sqlite --run content --dedupe 1

import sqlite3
import hashlib
import json
import os
import sys
import argparse
from pathlib import Path
import re


def clean_text(text: str) -> str:
    """Remove illegal surrogates and fix unicode issues"""
    return text.encode("utf-8", errors="ignore").decode("utf-8", errors="ignore")


def sha256_hex(data: bytes) -> str:
    """Calculate SHA-256 hash and return as hex string"""
    return hashlib.sha256(data).hexdigest()


def ensure_schema(conn: sqlite3.Connection):
    """Create database schema (FTS external content)"""
    cur = conn.cursor()

    cur.execute("PRAGMA journal_mode=WAL")
    cur.execute("PRAGMA synchronous=NORMAL")

    cur.execute("""
        CREATE TABLE IF NOT EXISTS docs(
            run_id TEXT NOT NULL,
            doc_id TEXT NOT NULL,
            meta_json TEXT,
            text_sha256 TEXT,
            text_chars INTEGER,
            PRIMARY KEY(run_id, doc_id)
        )
    """)

    # IMPORTANT: on remet text dans chunks (external content FTS ne duplique pas le texte)
    cur.execute("""
        CREATE TABLE IF NOT EXISTS chunks(
            id INTEGER PRIMARY KEY,
            uid TEXT NOT NULL UNIQUE,
            run_id TEXT NOT NULL,
            chunk_id TEXT NOT NULL,
            order_index INTEGER,
            doc_id TEXT,
            source_file TEXT,
            text TEXT
        )
    """)

    cur.execute("CREATE INDEX IF NOT EXISTS idx_chunks_doc ON chunks(run_id, doc_id, order_index)")
    cur.execute("CREATE INDEX IF NOT EXISTS idx_docs_hash ON docs(run_id, text_sha256)")
    cur.execute("CREATE UNIQUE INDEX IF NOT EXISTS idx_chunks_uid ON chunks(uid)")

    # FTS5 external content: le texte n'est pas stocké dans chunks_fts_content,
    # il référence chunks(text) via rowid=id
    cur.execute("""
        CREATE VIRTUAL TABLE IF NOT EXISTS chunks_fts USING fts5(
            text,
            content='chunks',
            content_rowid='id',
            tokenize="unicode61"
        )
    """)

    conn.commit()
    cur.close()


def parse_blocks(file_text: str) -> list:
    """Parse file into record blocks"""
    pattern = re.compile(r"^--- SOURCE:\s*(.+?)\s*---\s*$", re.MULTILINE)
    matches = list(pattern.finditer(file_text))

    if not matches:
        return []

    blocks = []
    for i, match in enumerate(matches):
        start = match.start()
        end = matches[i + 1].start() if i + 1 < len(matches) else len(file_text)
        header = match.group(1)

        chunk = file_text[start:end].strip()

        ms = re.search(r"METADATA_SOURCE:\s*(.+)", chunk)
        mf = re.search(r"METADATA_FILENAME:\s*(.+)", chunk)

        split_marker = "----------------------------------------"
        if split_marker in chunk:
            body = chunk.split(split_marker, 1)[1].strip()
        else:
            body = re.sub(r"^--- SOURCE:.*---\s*", "", chunk, flags=re.MULTILINE).strip()

        blocks.append({
            "source_line": header,
            "metadata_source": ms.group(1).strip() if ms else "",
            "metadata_filename": mf.group(1).strip() if mf else "",
            "body": body
        })

    return blocks


def extract_efta_id(s: str) -> str | None:
    """Extract EFTA ID from string"""
    match = re.search(r"\bEFTA\d{8}\b", s)
    return match.group(0) if match else None


def guess_type(body: str) -> str:
    """Guess if content is email-like or document-like"""
    head = body[:1500]
    looks_email = bool(
        re.search(r"(^|\n)\s*From:\s*", head, re.IGNORECASE)
        and re.search(r"(^|\n)\s*Subject:\s*", head, re.IGNORECASE)
    )
    return "email_like" if looks_email else "document_like"


def strip_noise(body: str) -> str:
    """Remove common OCR noise and boilerplate (UNCHANGED)"""
    t = body

    t = re.sub(r"--- PAGE \d+ ---", "\n", t)
    t = re.sub(r"^METADATA_(SOURCE|FILENAME):.*$", "", t, flags=re.MULTILINE | re.IGNORECASE)
    t = re.sub(r"^Couldn't load plugin\.\s*$", "", t, flags=re.MULTILINE | re.IGNORECASE)

    # OCR boilerplate
    t = re.sub(r"This communication contains information[\s\S]{0,900}?applicable law\.\s*", "", t, flags=re.IGNORECASE)
    t = re.sub(r"please note[\s\S]{0,1400}?copyright -all rights reserved\s*", "", t, flags=re.IGNORECASE)

    t = re.sub(r"\n{3,}", "\n\n", t).strip()
    return t


def chunk_paragraph(text: str, target_size: int = 2000) -> list[str]:
    """Split text into larger chunks (~target_size chars each) (UNCHANGED)"""
    parts = [p.strip() for p in re.split(r"\n\s*\n+", text) if p.strip()]

    chunks = []
    current = ""

    for p in parts:
        if current and len(current) + len(p) + 2 > target_size:
            chunks.append(current)
            current = p
        else:
            current = current + "\n\n" + p if current else p

    if current:
        chunks.append(current)

    return chunks if chunks else [text.strip()]


def supports_returning(conn: sqlite3.Connection) -> bool:
    """Detect whether SQLite likely supports RETURNING (>= 3.35)."""
    cur = conn.cursor()
    cur.execute("select sqlite_version()")
    v = cur.fetchone()[0]
    cur.close()

    try:
        major, minor, patch = (int(x) for x in v.split("."))
    except Exception:
        return False

    return (major, minor, patch) >= (3, 35, 0)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--input", required=True, help="Input directory")
    parser.add_argument("--db", default="records.sqlite", help="Database file")
    parser.add_argument("--run", default="content", help="Run ID")
    parser.add_argument("--dedupe", default="1", help="Enable deduplication (1 or 0)")
    parser.add_argument("--chunk-size", type=int, default=2000, help="Target chunk size in characters (default: 2000)")
    args = parser.parse_args()

    input_dir = Path(args.input)
    db_path = args.db
    run_id = args.run.strip()
    dedupe = args.dedupe == "1"
    chunk_size = args.chunk_size

    if not run_id:
        print("Error: --run must not be empty", file=sys.stderr)
        sys.exit(1)

    print(f'runId = "{run_id}" | dedupe={1 if dedupe else 0} | chunk_size={chunk_size}')

    conn = sqlite3.connect(db_path)
    conn.row_factory = sqlite3.Row

    try:
        ensure_schema(conn)
        has_returning = supports_returning(conn)
        print(f"SQLite RETURNING support: {has_returning}")

        # Test parameter binding
        cur = conn.cursor()
        cur.execute("SELECT ? as x", (run_id,))
        bind_test = cur.fetchone()[0]
        cur.close()
        if bind_test != run_id:
            raise Exception(f"Param binding broken: got {bind_test} expected {run_id}")

        doc_count = 0
        chunk_count = 0
        dup_skipped = 0

        cur = conn.cursor()

        # Prepared statements (faster + fewer surprises)
        stmt_doc = """
            INSERT INTO docs(run_id, doc_id, meta_json, text_sha256, text_chars)
            VALUES (?, ?, ?, ?, ?)
            ON CONFLICT(run_id, doc_id) DO UPDATE SET
                meta_json=excluded.meta_json,
                text_sha256=excluded.text_sha256,
                text_chars=excluded.text_chars
        """

        stmt_chunk_insert_returning = """
            INSERT INTO chunks(uid, run_id, chunk_id, order_index, doc_id, source_file, text)
            VALUES (?, ?, ?, ?, ?, ?, ?)
            ON CONFLICT(uid) DO NOTHING
            RETURNING id
        """

        stmt_chunk_insert_no_returning = """
            INSERT OR IGNORE INTO chunks(uid, run_id, chunk_id, order_index, doc_id, source_file, text)
            VALUES (?, ?, ?, ?, ?, ?, ?)
        """

        # External content FTS: on insère dans l'index (rowid = chunks.id)
        stmt_fts = """
            INSERT OR IGNORE INTO chunks_fts(rowid, text)
            VALUES (?, ?)
        """

        # Process all .txt files
        txt_files = list(input_dir.rglob("*.txt"))
        print(f"Found {len(txt_files)} .txt files to process\n")

        for file_idx, txt_file in enumerate(txt_files, 1):
            if file_idx % 10 == 0:
                print(f"Processing file {file_idx}/{len(txt_files)}: {txt_file.name}")

            try:
                file_text = clean_text(txt_file.read_text(encoding="utf-8", errors="ignore"))
            except Exception as e:
                print(f"Error reading {txt_file}: {e}", file=sys.stderr)
                continue

            blocks = parse_blocks(file_text)
            if not blocks:
                continue

            for block in blocks:
                # Extract or generate doc_id
                efta = (
                    extract_efta_id(block["source_line"])
                    or extract_efta_id(block["metadata_filename"])
                    or extract_efta_id(block["body"])
                )

                if efta:
                    doc_id = efta
                else:
                    hash_input = f"{block['source_line']}|{txt_file}"
                    doc_id = f"doc_{sha256_hex(hash_input.encode('utf-8'))}"

                body_clean = strip_noise(clean_text(block["body"]))
                text_sha = sha256_hex(body_clean.encode("utf-8"))

                # Check duplicates (UNCHANGED logic)
                if dedupe:
                    cur.execute(
                        "SELECT doc_id FROM docs WHERE run_id = ? AND text_sha256 = ? LIMIT 1",
                        (run_id, text_sha),
                    )
                    if cur.fetchone():
                        dup_skipped += 1
                        continue

                meta = {
                    "doc_id": doc_id,
                    "type": guess_type(body_clean),
                    "source_txt_file": str(txt_file),
                    "record_source": block["source_line"],
                    "metadata_source": block["metadata_source"],
                    "metadata_filename": block["metadata_filename"],
                    "text_sha256": text_sha,
                    "chars": len(body_clean),
                    "chunk_mode": "paragraph",
                    "chunk_target_size": chunk_size,
                }

                # Insert/update docs
                cur.execute(stmt_doc, (run_id, doc_id, json.dumps(meta), text_sha, len(body_clean)))
                doc_count += 1

                # Chunks
                parts = chunk_paragraph(body_clean, target_size=chunk_size)
                for i, part in enumerate(parts):
                    chunk_id = f"chunk_{str(chunk_count).zfill(8)}"
                    uid = f"{run_id}:{chunk_id}"

                    chunk_row_id = None

                    if has_returning:
                        cur.execute(
                            stmt_chunk_insert_returning,
                            (uid, run_id, chunk_id, i, doc_id, str(txt_file), part),
                        )
                        row = cur.fetchone()
                        # Si conflit uid => DO NOTHING => RETURNING renvoie None
                        if row:
                            chunk_row_id = row[0]
                    else:
                        # Fallback: insert OR IGNORE, puis lastrowid si insertion
                        cur.execute(
                            stmt_chunk_insert_no_returning,
                            (uid, run_id, chunk_id, i, doc_id, str(txt_file), part),
                        )
                        # rowcount = 1 si insert, 0 si ignore
                        if cur.rowcount == 1:
                            chunk_row_id = cur.lastrowid

                    if chunk_row_id is not None:
                        cur.execute(stmt_fts, (chunk_row_id, part))

                    chunk_count += 1

                # Commit periodically
                if doc_count % 200 == 0:
                    conn.commit()
                    conn.execute("PRAGMA wal_checkpoint(PASSIVE)")
                    print(f"docs: {doc_count} | chunks: {chunk_count} | dup_skipped: {dup_skipped}")

        conn.commit()
        cur.close()

        # Final checkpoint (reduce WAL)
        print("\nCheckpoint final du WAL...")
        conn.execute("PRAGMA wal_checkpoint(TRUNCATE)")
        conn.commit()

        # FTS optimize + PRAGMA optimize + ANALYZE + VACUUM
        # (VACUUM est le plus coûteux, mais c'est ce qui rend l'espace disque)
        print("Optimisation FTS et DB (optimize/analyze/vacuum)...")
        try:
            conn.execute("INSERT INTO chunks_fts(chunks_fts) VALUES('optimize')")
        except Exception as e:
            # Certains builds peuvent refuser optimize selon config; pas bloquant
            print(f"FTS optimize skipped: {e}", file=sys.stderr)

        try:
            conn.execute("PRAGMA optimize")
        except Exception:
            pass

        try:
            conn.execute("ANALYZE")
        except Exception:
            pass

        conn.commit()

        # Important: VACUUM après checkpoint + commit
        try:
            conn.execute("VACUUM")
        except Exception as e:
            print(f"VACUUM failed: {e}", file=sys.stderr)

        conn.commit()

        # Final stats
        cur = conn.cursor()
        cur.execute("SELECT COUNT(*) FROM docs WHERE run_id = ?", (run_id,))
        docs = cur.fetchone()[0]

        cur.execute("SELECT COUNT(*) FROM chunks WHERE run_id = ?", (run_id,))
        chunks = cur.fetchone()[0]

        # FTS entries: rows in chunks_fts
        cur.execute("SELECT COUNT(*) FROM chunks_fts")
        fts = cur.fetchone()[0]

        cur.execute("SELECT AVG(LENGTH(text)) FROM chunks WHERE run_id = ?", (run_id,))
        avg_chunk_size = cur.fetchone()[0] or 0

        cur.close()

        db_size_mb = os.path.getsize(db_path) / 1024 / 1024

        print(f"\n{'='*60}")
        print("✅ Done!")
        print(f"{'='*60}")
        print(f"Documents:           {docs:,}")
        print(f"Chunks:              {chunks:,}")
        print(f"FTS entries:         {fts:,}")
        print(f"Duplicates skipped:  {dup_skipped:,}")
        print(f"Avg chunk size:      {avg_chunk_size:.0f} chars")
        print(f"Database size:       {db_size_mb:.2f} MB")
        print(f"Database file:       {db_path}")
        print(f"{'='*60}")

    finally:
        conn.close()


if __name__ == "__main__":
    main()
